### PR TITLE
Add WASI support by tokio_wasi

### PR DESCRIPTION
**Pull request by langyo**
### Description

I've been trying to use `yew` to render the page into the static HTML string on WASI. However, `prokio` cannot distinguish the browser WASM target (`wasm32-unknown-unknown` with `wasm-bindgen`) and WASI target (`wasm32-wasi`), and it would choose wrong modules for `wasm32-*`.

This pull request consists of the following changes:

- Add the new runtime module which supported by `tokio_wasi` for `wasm32-wasi` target.

### Checklist

- [x] I have self-reviewed and tested this pull request.
- [x] I have added tests for my changes.
- [x] I have updated docs to reflect any new features / changes in this pull request.
- [x] ~~I have added my changes to `CHANGELOG.md`.~~

### CODE REVIEW

**Comment by ranile** [comment]
@futursolo Can you give this a review?

File: Cargo.toml [hunk 1]
```diff
@@ -12,22 +12,29 @@ repository = "https://github.com/futursolo/prokio"
...
+[target.'cfg(all(target_arch = "wasm32", not(target_os = "wasi")))'.dependencies]
+wasm-bindgen-futures = { version = "0.4" }
+gloo = { version = "0.10" }
 
-[target.'cfg(target_arch = "wasm32")'.dependencies]
-wasm-bindgen-futures = "0.4"
-gloo = "0.8"
+[target.'cfg(target_os = "wasi")'.dependencies]
+tokio_wasi = { version = "1", features = ["rt", "time"] }
+tokio-stream_wasi = { version = "0.1", features = ["time"] }
```
Comment (futursolo) [inline Cargo.toml]: I believe both rt and time are actually available in `tokio` for `wasm32-wasi` already.

Why do we need to use a fork of tokio here?

File: src/rt_tokio_wasi/mod.rs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        Self {
+            _marker: PhantomData,
+        }
+    }
+
+    pub fn spawn_local<F>(&self, f: F)
+    where
+        F: Future<Output = ()> + 'static,
+    {
+        RUNTIME.block_on(async { f.await })
```
Comment (futursolo) [inline src/rt_tokio_wasi/mod.rs]: To spawn a `!Send` future, you need to combine a LocalSet and a Runtime.
You can see how this is being done in the current tokio runtime.

`block_on` cannot be called in asynchronous execution context, which effectively prevents tasks from spawning.

File: src/rt_tokio_wasi/mod.rs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+use std::marker::PhantomData;
+
+use once_cell::sync::Lazy;
+
+static RUNTIME: Lazy<tokio::runtime::Runtime> = Lazy::new(|| {
+    tokio::runtime::Builder::new_current_thread()
+        .enable_all()
+        .build()
+        .unwrap()
+});
```
Comment (futursolo) [inline src/rt_tokio_wasi/mod.rs]: If a platform has no multi-threading support, i.e.: a thread runs on the main thread, the runtime should already be running via `#[tokio::main]`. Creating a different runtime assumed to be run in the same thread is not going to work as calling `block_on` effectively blocks the outer runtime.

The correct way to run a runtime that can execute a `!Send` future in the main thread is to call `tokio::task::LocalSet::run_until()` in `main()`.

See: https://docs.rs/tokio/latest/tokio/task/struct.LocalSet.html#method.run_until

**Review by futursolo — COMMENTED** [review]
Thank you for the pull request.

However, I feel we cannot support WASI as a spawnable runtime (`Runtime::new_...`) until multi-threading is properly supported as prokio requires the runtime to continuing resolve tasks spawned with `spawn_local()`. This requires either dedicated worker thread or a runtime that already runs for the lifetime of the program (`#[tokio::main(flavor = “current_thread")]` + `LocalSet`).

However, you can run `tokio::task::LocalSet::run_until()` in the `main()` function (See tokio documentation about [run_until](https://docs.rs/tokio/latest/tokio/task/struct.LocalSet.html#method.run_until)).
After that, if we add feature flags to redirect `wasm32-wasi` to use tokio instead of `wasm-bindgen`, prokio should work under WASI.

We can make `Runtime` a shim that redirects to `tokio::task::spawn_local`.

**Comment by langyo** [comment]
I did some research...Although I don’t quite understand your intention, we can indeed use `run_until` directly in a WASI program that already uses `#[tokio::main(flavor = “current_thread”)]`.

However, I created this PR to support `yew` being able to do SSR in a WASI environment. In order to do this, I modified many upstream libraries, including this repository.

When yew implements SSR rendering internally, it will create additional threads through prokio for node parsing. This seems to be a good way to speed up rendering, but doing so directly makes the original rendering logic unable to run in the WASI environment...After all, WASI has not yet incorporated multi-threading into the standard.

For implement the SSR rendering capability in WASI first, I currently added the function `render_async()` to [another PR](https://github.com/yewstack/yew/pull/3534) associated with this PR in `yew`. This function will render nodes serially without creating any additional asynchronous tasks.

The above function has achieved my goal and is already running in my current private business code. Maybe this PR needs to be modified to only ensure that `wasm_bindgen` is not introduced when compiling with WASI to avoid global symbol export pollution. I may only be able to avoid single-threaded rendering of asynchronous tasks through my new `render_async()` method before WASI officially supports multi-threading.

I've delete the incomplete WASI runtime implementation I implemented and continue to use the current `tokio` module. The only change currently by this PR is the correct handling of WASI compilation flags.

Umm...And don't forget my `yew`'s PR please...

cc @futursolo

**Comment by futursolo** [comment]
> When yew implements SSR rendering internally, it will create additional threads through prokio for node parsing. This seems to be a good way to speed up rendering, but doing so directly makes the original rendering logic unable to run in the WASI environment...After all, WASI has not yet incorporated multi-threading into the standard.

The reason why it is unable to do so is not because of lack of multi-threading but it assumes `wasm-bindgen-futures` instead of `tokio` for `wasm32` targets.

> This seems to be a good way to speed up rendering, but doing so directly makes the original rendering logic unable to run in the WASI environment...After all, WASI has not yet incorporated multi-threading into the standard.
> For implement the SSR rendering capability in WASI first, I currently added the function render_async() to https://github.com/yewstack/yew/pull/3534 associated with this PR in yew. This function will render nodes serially without creating any additional asynchronous tasks.

I intended to address `render_async` in the Yew pull request.
The purpose of the task spawning in `ServerRenderer` is not to speed up rendering but to convert a `!Send` Future to a `Send` one. Most tokio web frameworks require their futures to be `Send` and `wasm_bindgen` generated types are `!Send`, which makes Yew `!Send`.

You can see below what happens if you try to use the `_async` functions.

image


Instead of adding `_async` flavour, you should use `LocalServerRenderer` if you do not need `Send` futures and `ServerRender` if you need `Send` futures.

> I may only be able to avoid single-threaded rendering of asynchronous tasks through my new render_async() method before WASI officially supports multi-threading.

The `ServerRenderer` avoids single-threading by pinning the renderer to a worker thread that is least busy via `spawn_pinned`, which achieves multi threading when there is multiple worker threads in the prokio runtime.

**Comment by langyo** [comment]
> > When yew implements SSR rendering internally, it will create additional threads through prokio for node parsing. This seems to be a good way to speed up rendering, but doing so directly makes the original rendering logic unable to run in the WASI environment...After all, WASI has not yet incorporated multi-threading into the standard.
> 
> The reason why it is unable to do so is not because of lack of multi-threading but it assumes `wasm-bindgen-futures` instead of `tokio` for `wasm32` targets.
> 
> > This seems to be a good way to speed up rendering, but doing so directly makes the original rendering logic unable to run in the WASI environment...After all, WASI has not yet incorporated multi-threading into the standard.
> > For implement the SSR rendering capability in WASI first, I currently added the function render_async() to [yewstack/yew#3534](https://github.com/yewstack/yew/pull/3534) associated with this PR in yew. This function will render nodes serially without creating any additional asynchronous tasks.
> 
> I intended to address `render_async` in the Yew pull request. The purpose of the task spawning in `ServerRenderer` is not to speed up rendering but to convert a `!Send` Future to a `Send` one. Most tokio web frameworks require their futures to be `Send` and `wasm_bindgen` generated types are `!Send`, which makes Yew `!Send`.
> 
> You can see below what happens if you try to use the `_async` functions.
> 
> image
> Instead of adding `_async` flavour, you should use `LocalServerRenderer` if you do not need `Send` futures and `ServerRender` if you need `Send` futures.
> 
> > I may only be able to avoid single-threaded rendering of asynchronous tasks through my new render_async() method before WASI officially supports multi-threading.
> 
> The `ServerRenderer` avoids single-threading by pinning the renderer to a worker thread that is least busy via `spawn_pinned`, which achieves multi threading when there is multiple worker threads in the prokio runtime.

Thank you for your reminder. I didn't realize there was an interface called `LocalServerRenderer` until you gave me specific instructions.

I [made some modifications](https://github.com/langyo/yew/commit/133169e90ea8e9a2c4cb79e0f471bb2a8c92d897) based on your instructions. Now [there are no new interfaces](https://github.com/langyo/yew/commit/a520095c260b676d41b683d362bb62ddd323d41b), and the original WASI rendering example has been changed to use `LocalServerRenderer` for rendering.

Perhaps this interface should be specifically pointed out in the offical guide to remind latecomers... I've add some description [on the guide](https://github.com/langyo/yew/commit/b3f2b6e76a3639c7233668d5d73eb07947aa37fb).

File: Cargo.toml (hunk 1 above)
Comment (langyo) [reply Cargo.toml]: Solved at https://github.com/yewstack/prokio/pull/11/commits/f6ff913ff9fb9d5a179011e8b53ab1fdb704e273

File: Cargo.toml [hunk 4]
```diff
@@ -12,22 +12,27 @@ repository = "https://github.com/futursolo/prokio"
...
+
+[target.'cfg(all(target_arch = "wasm32", not(target_os = "wasi")))'.dependencies]
+wasm-bindgen-futures = { version = "0.4" }
+gloo = { version = "0.11" }
 
-[target.'cfg(target_arch = "wasm32")'.dependencies]
-wasm-bindgen-futures = "0.4"
-gloo = "0.8"
+[target.'cfg(any(not(target_arch = "wasm32"), target_os = "wasi"))'.dependencies]
+tokio = { version = "1", features = ["rt", "time"] }
```
Comment (futursolo) [inline Cargo.toml]:
```suggestion
tokio = { version = "1.21.0", features = ["rt", "time"] }
```

We should limit tokio to a version that has wasi support.

Comment (langyo) [reply Cargo.toml]: > We should limit tokio to a version that has wasi support.

Ummm... I tested it in my project. At least these capabilities are available in the latest version of `tokio`(1.35)...Is 1.21 too old?

Comment (futursolo) [reply Cargo.toml]: > This release is the first release of Tokio to intentionally support WASM. The sync,macros,io-util,rt,time features are stabilized on WASM. Additionally the wasm32-wasi target is given unstable support for the net feature.

> https://github.com/tokio-rs/tokio/blob/master/tokio/CHANGELOG.md#1210-september-2-2022

1.21.0 was based on the Changelog of tokio. But I am fine with a newer version as well.

Comment (langyo) [reply Cargo.toml]: > > This release is the first release of Tokio to intentionally support WASM. The sync,macros,io-util,rt,time features are stabilized on WASM. Additionally the wasm32-wasi target is given unstable support for the net feature.
> 
> > https://github.com/tokio-rs/tokio/blob/master/tokio/CHANGELOG.md#1210-september-2-2022
> 
> 1.21.0 was based on the Changelog of tokio. But I am fine with a newer version as well.

Done https://github.com/yewstack/prokio/pull/11/commits/5ba912c82fcbb0db040c43db38ce824e4fa7c7b8

**Review by futursolo — APPROVED** [review]
Thank you! Looks good to me.

The Runtime struct for tokio backend still needs some adjustments before it can properly support WASI without threading support, I will do this in a separate pull request.

After that, we can release prokio’s wasi support.

**Comment by langyo** [comment]
As #12 cannot be implemented for a while, could you release a patch version now to support `yew` to run normally in the WASI environment?

@futursolo

**Pull request merged**